user_management_service/app/storage/repos/WalletRepository.scala

```diff
@@ -0,0 +1,101 @@
+package storage.repos
+
+import exceptions.StorageException._
+import exceptions.StorageException.WalletStorageException._
+import storage.db.WalletTableRepository
+import storage.model.WalletResource
+
+import akka.Done
+import org.postgresql.util.PSQLException
+import play.api.db.slick.{DatabaseConfigProvider, HasDatabaseConfigProvider}
+import slick.jdbc.JdbcProfile
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class WalletRepository @Inject() (
+      protected val dbConfigProvider: DatabaseConfigProvider,
+      val tableRepository: WalletTableRepository
+    )(implicit ec: ExecutionContext)
+      extends HasDatabaseConfigProvider[JdbcProfile] {
+  import profile.api._
+  import tableRepository._
+
+  def get(userId: Long): Future[Option[WalletResource]] = db.run {
+    wallets.filter(_.userId === userId).result.headOption
+  }
+
+  def create(walletResource: WalletResource): Future[WalletResource] =
+    db.run(wallets += walletResource).map(_ => walletResource).recoverWith {
+      case e: PSQLException =>
+        Future.failed(UnknownDatabaseError(cause = Some(e)))
+    }
+
+  def transfer(
```

Comment:
I'm afraid this operation is not safe.
First of all, it should be transactional, so it would be rolled back if anything happens in between.
Second, you even may want to lock particular row with `select ... for update`, so if any two operations on the same balance occurs in parallel, you would get sequential changes with actual state.

Reply:
Also, additional validations could have place, like do nothing and fail operations if balance is less than required for that operation.

Reply:
Should such validation take place here or in resource handler is another question.

Reply:
I think transferAction should meet all requirements now. Marking as resolved